<a href="https://colab.research.google.com/github/EngRenan/Covid_Vaccination/blob/main/Vacinacao_Covid19.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Analise e coleta de dados provenientes do Kaggle a respeito de estatisticas das vacinas contra a pandemia de COVID 

In [8]:
# Importando bibliotecas
import pandas as pd
import os

In [9]:
# Importando dados do arquivo
Diretorio = '/content/'
arquivo = 'country_vaccinations.csv'
dados = os.path.join(Diretorio, arquivo) # Busca caminho do arquivo com suas conversões

In [10]:
# Verifica o caminho do diretorio que o arquivo esta salvo
dados

'/content/country_vaccinations.csv'

Agora sera criado um data frame no pandas e ler o arquivo na pasta de origem. Para abrir o arquivo vamos utilizar o comando read_csv e utilizar o separador de ',' de acordo com o dado fornecido e tambem utilizar o encoding informando a codificação do arquivo

In [11]:
dataframe = pd.read_csv(dados, sep=',', encoding='latin-1')

In [12]:
# Exibindo os primeiros registros da tabela para verificação
dataframe.head(5)

country iso_code        date  total_vaccinations  people_vaccinated  \
0  Afghanistan      AFG  2021-02-22                 0.0                0.0   
1  Afghanistan      AFG  2021-02-23                 NaN                NaN   
2  Afghanistan      AFG  2021-02-24                 NaN                NaN   
3  Afghanistan      AFG  2021-02-25                 NaN                NaN   
4  Afghanistan      AFG  2021-02-26                 NaN                NaN   

   people_fully_vaccinated  daily_vaccinations_raw  daily_vaccinations  \
0                      NaN                     NaN                 NaN   
1                      NaN                     NaN              1367.0   
2                      NaN                     NaN              1367.0   
3                      NaN                     NaN              1367.0   
4                      NaN                     NaN              1367.0   

   total_vaccinations_per_hundred  people_vaccinated_per_hundred  \
0                             0.0                            0.0   
1                             NaN                            NaN   
2                             NaN                            NaN   
3                             NaN                            NaN   
4                             NaN                            NaN   

   people_fully_vaccinated_per_hundred  daily_vaccinations_per_million  \
0                                  NaN                             NaN   
1                                  NaN                            34.0   
2                                  NaN                            34.0   
3                                  NaN                            34.0   
4                                  NaN                            34.0   

                                            vaccines  \
0  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
1  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
2  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
3  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
4  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   

                 source_name            source_website  
0  World Health Organization  https://covid19.who.int/  
1  World Health Organization  https://covid19.who.int/  
2  World Health Organization  https://covid19.who.int/  
3  World Health Organization  https://covid19.who.int/  
4  World Health Organization  https://covid19.who.int/

# Analisando os dados

Para realizar a analise vamos responder algumas perguntas: 

1. Quais sao os paises com maior numero de pessoas imunizadas? 
2. Qual vacina é a mais utilizada no mundo?
3. Qual país possui o maior numero de vacinados por dia? 
4. Qual a media de vacinação por país? 

In [13]:
# Verificando dados referente ao Brasil por dia (teste)
dataframe.query('country == "Brazil" ').sort_values('date', ascending=False)[:5]

country iso_code        date  total_vaccinations  people_vaccinated  \
11085  Brazil      BRA  2022-03-29         413559595.0        181078067.0   
11084  Brazil      BRA  2022-03-28         412177025.0        180881628.0   
11083  Brazil      BRA  2022-03-27         411548202.0        180800024.0   
11082  Brazil      BRA  2022-03-26         411116197.0        180757995.0   
11081  Brazil      BRA  2022-03-25         410860641.0        180726798.0   

       people_fully_vaccinated  daily_vaccinations_raw  daily_vaccinations  \
11085              160272858.0               1382570.0            656803.0   
11084              160008924.0                628823.0            546999.0   
11083              159842858.0                432005.0            587488.0   
11082              159756178.0                255556.0            548819.0   
11081              159673499.0                684805.0            596456.0   

       total_vaccinations_per_hundred  people_vaccinated_per_hundred  \
11085                          193.26                          84.62   
11084                          192.61                          84.53   
11083                          192.32                          84.49   
11082                          192.12                          84.47   
11081                          192.00                          84.45   

       people_fully_vaccinated_per_hundred  daily_vaccinations_per_million  \
11085                                74.90                          3069.0   
11084                                74.77                          2556.0   
11083                                74.70                          2745.0   
11082                                74.65                          2565.0   
11081                                74.62                          2787.0   

                                                vaccines  \
11085  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
11084  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
11083  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
11082  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
11081  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   

                                           source_name  \
11085  State governments via coronavirusbra1.github.io   
11084  State governments via coronavirusbra1.github.io   
11083  State governments via coronavirusbra1.github.io   
11082  State governments via coronavirusbra1.github.io   
11081  State governments via coronavirusbra1.github.io   

                          source_website  
11085  https://coronavirusbra1.github.io  
11084  https://coronavirusbra1.github.io  
11083  https://coronavirusbra1.github.io  
11082  https://coronavirusbra1.github.io  
11081  https://coronavirusbra1.github.io

In [14]:
# Verificando o numero de paises distintos na base de dados
dataframe['country'].nunique()

223

Observado que temos registro de 223 paises no dataset

# 1. Qual o país com maior numero de pessoas imunizadas? 

Selecionaremos o ultimo registro (soma total) de cada país realizando um agrupamento do país e selecionando a ultima data de atualização que foi realizada

In [15]:
dataframe.groupby('country')['date'].max().sort_values(ascending=False)

country
Lebanon                            2022-03-29
Lithuania                          2022-03-29
Kosovo                             2022-03-29
Kazakhstan                         2022-03-29
Japan                              2022-03-29
                                      ...    
Pitcairn                           2021-09-07
Turkmenistan                       2021-09-05
Bonaire Sint Eustatius and Saba    2021-09-01
Saint Helena                       2021-05-05
Falkland Islands                   2021-04-14
Name: date, Length: 223, dtype: object

Visualizado que existem paises qie nao possuem a atualização da ultima data da coleta. Dessa forma iremos utilizar a varivavel total_vaccinations no qual contem o total de vacinações naquele país. Podemos usar como parametro para buscar a ultima data de coleta. Dessa forma é necessario agrupar por país e pegar o valor máximo obtido nessa variável.

In [16]:
dataframe.groupby('country')['total_vaccinations'].max().sort_values(ascending=False)

country
China               3.263129e+09
India               1.834501e+09
United States       5.601818e+08
Brazil              4.135596e+08
Indonesia           3.771089e+08
                        ...     
Falkland Islands    4.407000e+03
Montserrat          4.211000e+03
Niue                4.161000e+03
Tokelau             1.936000e+03
Pitcairn            9.400000e+01
Name: total_vaccinations, Length: 223, dtype: float64

In [17]:
# Transformando em serie Pandas
imunizados_pais = dataframe.groupby('country')['total_vaccinations'].max().sort_values(ascending=False)
type(imunizados_pais)

pandas.core.series.Series

In [18]:
# Transformando em Dataframe novamente e resetar o index
imunizados = imunizados_pais.to_frame().reset_index()
imunizados

country  total_vaccinations
0               China        3.263129e+09
1               India        1.834501e+09
2       United States        5.601818e+08
3              Brazil        4.135596e+08
4           Indonesia        3.771089e+08
..                ...                 ...
218  Falkland Islands        4.407000e+03
219        Montserrat        4.211000e+03
220              Niue        4.161000e+03
221           Tokelau        1.936000e+03
222          Pitcairn        9.400000e+01

[223 rows x 2 columns]

In [19]:
# Criando lista dos top 10 paises 
imunizados.sort_values('total_vaccinations',ascending=False).head(10)

country  total_vaccinations
0          China        3.263129e+09
1          India        1.834501e+09
2  United States        5.601818e+08
3         Brazil        4.135596e+08
4      Indonesia        3.771089e+08
5          Japan        2.543456e+08
6     Bangladesh        2.436427e+08
7       Pakistan        2.193686e+08
8        Vietnam        2.031444e+08
9         Mexico        1.919079e+08

# 2. Qual a vacina mais utilizada no mundo?

Iremos criar um dataframe contendo as vacinas utilizadas por país. Com isso teremos uma relação com todas as vacinas por país em uma determinada data. 

In [20]:
vacinas_por_pais = dataframe[['country', 'date', 'vaccines']]
vacinas_por_pais

country        date  \
0      Afghanistan  2021-02-22   
1      Afghanistan  2021-02-23   
2      Afghanistan  2021-02-24   
3      Afghanistan  2021-02-25   
4      Afghanistan  2021-02-26   
...            ...         ...   
86507     Zimbabwe  2022-03-25   
86508     Zimbabwe  2022-03-26   
86509     Zimbabwe  2022-03-27   
86510     Zimbabwe  2022-03-28   
86511     Zimbabwe  2022-03-29   

                                                vaccines  
0      Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...  
1      Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...  
2      Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...  
3      Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...  
4      Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...  
...                                                  ...  
86507  Oxford/AstraZeneca, Sinopharm/Beijing, Sinovac...  
86508  Oxford/AstraZeneca, Sinopharm/Beijing, Sinovac...  
86509  Oxford/AstraZeneca, Sinopharm/Beijing, Sinovac...  
86510  Oxford/AstraZeneca, Sinopharm/Beijing, Sinovac...  
86511  Oxford/AstraZeneca, Sinopharm/Beijing, Sinovac...  

[86512 rows x 3 columns]

In [21]:
# Agrapando as vacinas e contando o país apenas uma vez. Se no dataset tiver um país utilizando duas vacinas, será contabilizado duas vezes. 
vacinas_por_pais.groupby('vaccines')['country'].nunique().sort_values(ascending=False)

vaccines
Oxford/AstraZeneca                                                                                    20
Johnson&Johnson, Moderna, Oxford/AstraZeneca, Pfizer/BioNTech                                         17
Moderna, Oxford/AstraZeneca, Pfizer/BioNTech                                                          15
Oxford/AstraZeneca, Pfizer/BioNTech                                                                   11
Johnson&Johnson, Moderna, Novavax, Oxford/AstraZeneca, Pfizer/BioNTech                                 8
                                                                                                      ..
Moderna, Oxford/AstraZeneca, Pfizer/BioNTech, Sinopharm/Beijing, Sinovac                               1
Moderna, Oxford/AstraZeneca, Pfizer/BioNTech, Sinovac, Sputnik Light, Sputnik V, ZF2001                1
Moderna, Oxford/AstraZeneca, Pfizer/BioNTech, Sinovac, Sputnik V                                       1
Moderna, Oxford/AstraZeneca, Pfizer/BioNTech, 

In [22]:
vacinas_pais = dataframe.groupby('country')['vaccines'].unique()
vacinas_pais = vacinas_pais.to_frame().reset_index()

Observado que Oxford/AstraZeneca é a vacina mais utilizada pelos paises  

# 3. Qual país possui o maior numero de vacinados por dia?

In [23]:
# Iremos avaliar a variavel daily_vaccinations
dataframe.groupby('country')['daily_vaccinations'].max().sort_values(ascending=False)

country
China            22424286.0
India            10037995.0
Bangladesh        3758404.0
United States     3506960.0
Pakistan          2175773.0
                    ...    
Saint Helena          165.0
Niue                   87.0
Montserrat             53.0
Tokelau                23.0
Pitcairn                1.0
Name: daily_vaccinations, Length: 223, dtype: float64

Analisado e visto que China, India, Bangladesh e United States tem o maior numero de vacinados por dia. 

# Qual a media de vacinação por país?

In [24]:
media_pais = dataframe.groupby('country')['daily_vaccinations'].mean().sort_values(ascending=False)
media_pais = media_pais.to_frame().reset_index()
media_pais

country  daily_vaccinations
0                China        6.930368e+06
1                India        4.175994e+06
2        United States        1.191727e+06
3               Brazil        9.435287e+05
4            Indonesia        8.462893e+05
..                 ...                 ...
218  Wallis and Futuna        3.388649e+01
219            Tokelau        1.739823e+01
220               Niue        1.631698e+01
221         Montserrat        1.053659e+01
222           Pitcairn        5.119048e-01

[223 rows x 2 columns]

Paises que tem a maior media de vacinação sao: China, India, USA, Brazil e Indonesia

# Criando base de dados somente com os dados tratados anteriormente

In [25]:
# Renomeando colunas 
media_pais = media_pais.rename(columns={'country': 'pais', 'daily_vaccinations': 'media_vacinacao'})
media_pais

pais  media_vacinacao
0                China     6.930368e+06
1                India     4.175994e+06
2        United States     1.191727e+06
3               Brazil     9.435287e+05
4            Indonesia     8.462893e+05
..                 ...              ...
218  Wallis and Futuna     3.388649e+01
219            Tokelau     1.739823e+01
220               Niue     1.631698e+01
221         Montserrat     1.053659e+01
222           Pitcairn     5.119048e-01

[223 rows x 2 columns]

In [28]:
vacinas_pais = vacinas_pais.rename(columns={'country': 'pais', 'vaccines': 'vacina_utilizada'})
vacinas_pais[:5]

pais                                   vacina_utilizada
0  Afghanistan  [Johnson&Johnson, Oxford/AstraZeneca, Pfizer/B...
1      Albania  [Oxford/AstraZeneca, Pfizer/BioNTech, Sinovac,...
2      Algeria  [Oxford/AstraZeneca, Sinopharm/Beijing, Sinova...
3      Andorra     [Moderna, Oxford/AstraZeneca, Pfizer/BioNTech]
4       Angola                               [Oxford/AstraZeneca]

In [27]:
imunizados = imunizados.rename(columns={'country': 'pais', 'total_vaccinations': 'Imunizados'})
imunizados[:5]

pais    Imunizados
0          China  3.263129e+09
1          India  1.834501e+09
2  United States  5.601818e+08
3         Brazil  4.135596e+08
4      Indonesia  3.771089e+08

# Criando novo dataset

Para criar um novo dataset sera necessário utilizar a função merge no qual vai comparar os campos das tabelas que são chaves iguais, como o nome do país e realizar uma junção entre as tabelas, lendo todos os dados tratados em uma unica tabela. 

In [31]:
# Realizando merge
dados_tratados = pd.merge(imunizados, media_pais, how='inner')

dados_tratados = pd.merge(dados_tratados, vacinas_pais)
dados_tratados[:10]

pais    Imunizados  media_vacinacao  \
0          China  3.263129e+09     6.930368e+06   
1          India  1.834501e+09     4.175994e+06   
2  United States  5.601818e+08     1.191727e+06   
3         Brazil  4.135596e+08     9.435287e+05   
4      Indonesia  3.771089e+08     8.462893e+05   
5          Japan  2.543456e+08     6.215795e+05   
6     Bangladesh  2.436427e+08     5.453055e+05   
7       Pakistan  2.193686e+08     5.430051e+05   
8        Vietnam  2.031444e+08     5.310949e+05   
9         Mexico  1.919079e+08     4.134253e+05   

                                    vacina_utilizada  
0  [CanSino, Sinopharm/Beijing, Sinopharm/Wuhan, ...  
1           [Covaxin, Oxford/AstraZeneca, Sputnik V]  
2        [Johnson&Johnson, Moderna, Pfizer/BioNTech]  
3  [Johnson&Johnson, Oxford/AstraZeneca, Pfizer/B...  
4  [Johnson&Johnson, Moderna, Novavax, Oxford/Ast...  
5     [Moderna, Oxford/AstraZeneca, Pfizer/BioNTech]  
6  [Johnson&Johnson, Moderna, Oxford/AstraZeneca,...  
7  [CanSino, Covaxin, Moderna, Oxford/AstraZeneca...  
8  [Abdala, Moderna, Oxford/AstraZeneca, Pfizer/B...  
9  [CanSino, Johnson&Johnson, Moderna, Oxford/Ast...

# Exportando dados tratados

In [32]:
dados_tratados.to_csv('estatistica_pais.csv', index=True, sep=';')

In [33]:
dados_tratados.to_json('estatistica.json', orient='records')